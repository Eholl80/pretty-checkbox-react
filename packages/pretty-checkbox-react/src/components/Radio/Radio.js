// @flow

import * as React from 'react';
import classNames from 'classnames';

import { mergeRefs } from '../../utils/utils';
import { BaseInput } from '../BaseInput/BaseInput';

import {
    type PCRColors,
    type PCRColorsOutline,
    type RadioCheckboxVariants,
    type RadioCheckboxFillModes,
    type RadioCheckboxAnimations,
    type IconPropType,
} from '../../types';

export type RadioState = boolean | string;

export type RadioProps = {
    /**
     * Specify one of the colors.
     */
    color?: PCRColors | PCRColorsOutline,

    /**
     * Specify one of the variants.
     */
    variant?: RadioCheckboxVariants,

    /**
     * Specify one of the fill modes.
     */
    fill?: RadioCheckboxFillModes,

    /**
     * Specify one of the animations.
     */
    animation?: RadioCheckboxAnimations,

    /**
     * Controlled mode checkbox.
     */
    checked?: boolean,

    /**
     * Pass an icon to render.
     */
    icon?: IconPropType,

    /**
     * Runs inside the default onChange handler for the Checkbox.
     *
     * Optionally return `false` from a synchronous action to prevent the
     * checkbox from being un/checked.
     */
    onChange?: (e: SyntheticMouseEvent<HTMLInputElement>) => ?boolean,

    /**
     * Set the checkbox as disabled.
     */
    disabled?: boolean,

    /**
     * Set the checkbox as locked. This will also set the
     * `aria-disabled` attribute to `true`.
     */
    locked?: boolean,

    /**
     * Pass additional class names to the `.pretty` div.
     */
    className?: string,

    /**
     * The `state` prop from` useRadioState`.
     */
    state?: RadioState,

    /**
     * The `setState` dispatcher from `useRadioState`.
     */
    setState?: (state: RadioState) => void,

    /**
     * Stub...
     */
    value?: string,

    /**
     * Additional HTML attributes to pass to the `input` element.
     * TODO: fix flow type after https://github.com/facebook/flow/pull/7569 ðŸ™„
     */
    htmlProps?: Object,

    /**
     * Text to show inside the label _or_ a render function
     * that allows you to render the internal pretty-checkbox
     * `.state` div.
     *
     * @example
     * <Radio>
     *     {({ color, icon }) => (
     *         <div className="state my-custom-class-name">
     *             {icon}
     *             <label>Something super special</label>
     *         </div>
     *     )}
     * </Radio>
     */
    children: React.Node | ((...props: Object) => React.Node),
};

export const useRadioState = (initialState: RadioState | (() => RadioState)) => {
    const [state, setState] = React.useState<RadioState>(initialState);

    return { state, setState };
};

export const Radio = React.forwardRef<RadioProps, HTMLInputElement>(({
    color,
    variant = 'round',
    fill,
    animation,
    checked,
    onChange,
    disabled,
    locked,
    children,
    className,
    icon = { type: null, icon: null },
    value,
    state,
    setState,
    ...htmlProps
}, ref) => {
    const radioRef = React.useRef(mergeRefs(ref, React.createRef()));

    const handleChange = React.useCallback(
        (e: SyntheticMouseEvent<HTMLInputElement>) => {
            if (disabled || locked) {
                e.preventDefault();
                return;
            }

            if (typeof onChange === 'function') {
                if (onChange(e) === false) {
                    return;
                }
            }

            if (typeof setState !== 'function') {
                return;
            }

            setState(e.currentTarget.value);
        },
        [setState, state, disabled, locked, onChange]
    );

    return (
        <BaseInput
            type="radio"
            role="radio"
            locked={locked}
            icon={icon}
            className={classNames(
                {
                    'p-default': !icon.type && (!animation || animation === 'smooth'),
                    [`p-${variant}`]: variant,
                    [animation ? `p-${animation}` : '']: animation,
                    [icon.type ? `p-${icon.type}` : '']: icon.type,
                    [fill ? `p-${fill}` : '']: fill,
                },
                className
            )}
            disabled={disabled}
            aria-checked={!!checked}
            aria-disabled={disabled || locked}
            // $FlowFixMe
            onChange={handleChange}
            checked={checked || (state && state === value)}
            value={value}
            tabIndex="0"
            ref={radioRef.current}
            {...htmlProps}>
            {children}
        </BaseInput>
    );
});

export const RadioGroup = (props: Object) => <fieldset role="radiogroup" {...props} />;
