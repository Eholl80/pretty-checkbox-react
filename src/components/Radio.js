// @flow

import React from 'react';
import classNames from 'classnames';
import { getBaseClassName } from '../utils/utils';

import Input, { PREFIX } from './Input';
import type { InputProps } from './Input';

export type RadioProps = {
    ...InputProps,

    /**
     * The name of the radio group.
     */
    name: string
};

function Radio(props: RadioProps) {
    const { className, name, inputProps, ...rest } = props;

    return (
        <Input
            type="radio"
            className={
                classNames(
                    // $ExpectError
                    getBaseClassName(props, PREFIX),
                    className
                )
            }
            inputProps={{ ...inputProps, name: name }}
            {...rest}
        />
    );
}

Radio.defaultProps = { shape: 'round' };

export default Radio;
