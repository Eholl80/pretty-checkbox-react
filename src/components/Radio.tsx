import * as React from 'react';

import { useIcon } from '../hooks/useIcon';
import { CommonCheckboxRadioProps } from '../types/CommonProps';
import { Pretty } from '../factory/Pretty';
import { isBoolean } from '../factory/utils/utils';
import { useUUID } from '../hooks/useUUID';
import { Group, GroupProps } from './Group';

export type RadioState = boolean | string;
export type RadioProps = CommonCheckboxRadioProps<RadioState>;

export const useRadioState = ({ state: initialState = false }: { state?: RadioState } = {}) => {
    const [state, setState] = React.useState<RadioState>(initialState);

    return {
        state,
        setState,
        onChange: React.useCallback((e: React.ChangeEvent<HTMLInputElement>, args?: RadioState) => {
            const value = args || e.currentTarget.value;
            setState(prev => (isBoolean(prev) ? !prev : value));
        }, []),
        ...useUUID(),
    };
};

export const RadioGroup = (props: GroupProps) => (
    <Group as="fieldset" role="radiogroup" {...props} />
);

export const Radio = React.forwardRef<HTMLInputElement, RadioProps>(
    ({ value: userValue, icon: userIcon, shape = 'round', ...rest }: RadioProps, ref) => {
        const { icon, iconType } = useIcon(userIcon);
        const value = typeof userValue === 'undefined' ? '' : userValue;

        return React.createElement(Pretty, {
            type: 'radio',
            shape,
            icon,
            iconType,
            value,
            ref,
            ...rest,
        });
    }
);

Radio.displayName = 'Radio';
