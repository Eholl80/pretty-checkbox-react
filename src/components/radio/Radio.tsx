import * as React from 'react';
import classNames from 'classnames';

import { State } from '../state/State';

import { useLocked } from '../../hooks/utility/useLocked';
import { useIcon } from '../../hooks/useIcon';
import { useClassNames } from '../../hooks/utility/useClassNames';
import { useControlled } from '../../hooks/utility/useControlled';

import { useCheckboxRadioProps } from '../../hooks/utility/useCheckboxRadioProps';
import { PCRCheckboxRadioProps } from '../../typings/PCRCheckboxRadioProps';
import { UseRadioState, useRadioState } from './useRadioState';

export const Radio = React.forwardRef<
    HTMLInputElement,
    PCRCheckboxRadioProps<UseRadioState['state']>
>((props, ref) => {
    const { checked, value, ...rest } = useControlled<
        PCRCheckboxRadioProps<UseRadioState['state']>
    >(props);

    const {
        shape = 'round',
        children,
        locked,
        color,
        id,
        className,
        style,
        icon: propsIcon,
        htmlProps,
    } = useCheckboxRadioProps(rest);

    const styles = useLocked({ locked, style });
    const { icon, iconType } = useIcon(propsIcon);

    return (
        <div
            style={styles}
            className={classNames(
                'pretty',
                useClassNames({
                    ...props,
                    shape,
                    iconType,
                }),
                className
            )}>
            <input ref={ref} value={value} type="radio" id={id} checked={checked} {...htmlProps} />
            <State id={id} icon={icon} color={color}>
                {children}
            </State>
        </div>
    );
});

Radio.displayName = 'Radio';

export { UseRadioState, useRadioState };
