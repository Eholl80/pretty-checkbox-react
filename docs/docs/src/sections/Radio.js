import React from 'react';

import { Radio, Switch } from 'pretty-checkbox-react';

import { Title } from '../components/Title';
import { CollapseContainer } from '../components/CollapseContainer';

import {
    LiveProvider,
    LiveEditor,
    LiveError,
    LivePreview
} from 'react-live';

const scope = { Radio, Switch };

const basicExample = `<>
    <Radio name="radio-1">Same Day</Radio>
    <Radio name="radio-1">Next Dat</Radio>
    <Radio name="radio-1">Standard Ground</Radio>
</>`;

const colorsExample = `<>
    <Radio color="primary-o" shape="curve" name="radio-2">Primary</Radio>
    <Radio color="success-o" shape="curve" name="radio-2">Success</Radio>
    <Radio color="info-o" shape="curve" name="radio-2">Info</Radio>
    <Radio color="warning-o" shape="curve" name="radio-2">Warning</Radio>
    <Radio color="danger-o" shape="curve" name="radio-2">Danger</Radio>
</>`;

const solidColorsAndIconsExample = `<>
    <Radio color="primary" icon={<i className="mdi mdi-check" />} name="radio-3">New York</Radio>
    <Radio color="success" icon={<i className="mdi mdi-check" />} name="radio-3">Paris</Radio>
    <Radio color="info" icon={<i className="mdi mdi-check" />} name="radio-3">Dubai</Radio>
    <Radio color="warning" icon={<i className="mdi mdi-check" />} name="radio-3">Tokyo</Radio>
    <Radio color="danger" icon={<i className="mdi mdi-check" />} name="radio-3">Istanbul</Radio>
</>`;

const outlineColorsExample = `<>
    <Radio color="primary-o" icon={<i className="mdi mdi-check" />} name="radio-4">Dollar</Radio>
    <Radio color="success-o" icon={<i className="mdi mdi-check" />} name="radio-4">Euro</Radio>
    <Radio color="info-o" icon={<i className="mdi mdi-check" />} name="radio-4">Dinar</Radio>
    <Radio color="warning-o" icon={<i className="mdi mdi-check" />} name="radio-4">Pound</Radio>
    <Radio color="danger-o" icon={<i className="mdi mdi-check" />} name="radio-4">Rupee</Radio>
</>`;

const animationExample = `<>
    <Radio color="primary-o"
        shape="curve"
        animation="tada"
        icon={<i className="mdi mdi-check" />}
        name="radio-5">HTML</Radio>

    <Radio color="success-o"
        shape="curve"
        animation="rotate"
        icon={<i className="mdi mdi-check" />}
        name="radio-5">JavaScript</Radio>

    <Radio color="info-o"
        shape="curve"
        animation="pulse"
        icon={<i className="mdi mdi-check" />}
        name="radio-5">Python</Radio>

    <Radio color="warning-o"
        shape="curve"
        animation="jelly"
        icon={<i className="mdi mdi-check" />}
        name="radio-5">PHP</Radio>

    <Radio color="danger-o"
        shape="curve"
        animation="smooth"
        icon={<i className="mdi mdi-check" />}
        name="radio-5">Swift</Radio>
</>`;

const plainExample = `<>
    <Radio plain
        color="primary-o"
        animation="smooth"
        icon={<i className="mdi mdi-human" />}
        name="radio=6">Single</Radio>

    <Radio plain
        color="info-o"
        animation="smooth"
        icon={<i className="mdi mdi-human-male-female" />}
        name="radio=6">Married</Radio>

    <Radio plain
        color="success-o"
        animation="smooth"
        icon={<i className="mdi mdi-heart" />}
        name="radio=6">In a relationship</Radio>
</>`;

const switchExample = `<>
    <Switch type="radio" color="success" name="radio-7">Summer</Switch>
    <Switch type="radio" color="success" style="fill" name="radio-7">Winter</Switch>
    <Switch type="radio" color="success" style="slim" name="radio-7">Fall</Switch>
</>`;

function PrettyRadio() {
    return (
        <>
            <Title>Radio</Title>
            <div className="content">
                <p>Styling radio buttons are very similar to checkbox. All features mentioned above will work for radio buttons as well!</p>
                <LiveProvider code={basicExample} scope={scope}>
                    <CollapseContainer title="Basic" demo={<LivePreview />}>
                        <>
                            <LiveEditor />
                            <LiveError />
                        </>
                    </CollapseContainer>
                </LiveProvider>
                <LiveProvider code={colorsExample} scope={scope}>
                    <CollapseContainer title="Colors" demo={<LivePreview />}>
                        <>
                            <LiveEditor />
                            <LiveError />
                        </>
                    </CollapseContainer>
                </LiveProvider>
                <LiveProvider code={solidColorsAndIconsExample} scope={scope}>
                    <CollapseContainer title="Solid color and icons" demo={<LivePreview />}>
                        <>
                            <LiveEditor />
                            <LiveError />
                        </>
                    </CollapseContainer>
                </LiveProvider>
                <LiveProvider code={outlineColorsExample} scope={scope}>
                    <CollapseContainer title="Outline colors" demo={<LivePreview />}>
                        <>
                            <LiveEditor />
                            <LiveError />
                        </>
                    </CollapseContainer>
                </LiveProvider>
                <LiveProvider code={animationExample} scope={scope}>
                    <CollapseContainer title="Animations" demo={<LivePreview />}>
                        <>
                            <LiveEditor />
                            <LiveError />
                        </>
                    </CollapseContainer>
                </LiveProvider>
                <LiveProvider code={plainExample} scope={scope}>
                    <CollapseContainer title="Plain" demo={<LivePreview />}>
                        <>
                            <LiveEditor />
                            <LiveError />
                        </>
                    </CollapseContainer>
                </LiveProvider>
                <LiveProvider code={switchExample} scope={scope}>
                    <CollapseContainer title="Switch" demo={<LivePreview />}>
                        <>
                            <LiveEditor />
                            <LiveError />
                        </>
                    </CollapseContainer>
                </LiveProvider>
            </div>
        </>
    );
}

export default PrettyRadio;
