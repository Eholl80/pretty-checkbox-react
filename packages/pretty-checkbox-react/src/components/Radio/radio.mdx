---
name: Radio
---

import { useRadioState, Radio, RadioGroup } from './Radio.js';
import { Playground, Props } from 'docz';

# Radio

<Props of={Radio} />

## Basic Usage with useRadioState

<Playground>
    {() => {
        const radio = useRadioState();

        return (
            <>
                <Radio {...radio} value="apple" name="fruit">Apple</Radio>
                <Radio {...radio} value="orange" name="fruit">Orange</Radio>
                <Radio {...radio} value="watermelon" name="fruit">Watermelon</Radio>
            </>
        );
    }}
</Playground>

### Using without the HTML `name` attribute
Previously in PCR 1.0 the `name` prop was required. Well this is no more! Radio groups will
behave as expected with or without the name prop.

<Playground>
    {() => {
        const radio = useRadioState();

        return (
            <>
                <Radio {...radio} value="apple">Apple</Radio>
                <Radio {...radio} value="orange">Orange</Radio>
                <Radio {...radio} value="watermelon">Watermelon</Radio>
            </>
        );
    }}
</Playground>

## Basic usage without Hooks
Similar to Checkbox, Radio can be used using a `class`.

## Disabled and Locked Radio

<Playground>
    <>
        <Radio disabled>Disabled</Radio>
        <Radio locked>Locked</Radio>
    </>
</Playground>

## RadioGroup
PCR 2.0 introduces a new component called `RadioGroup`.
It's just a wrapper around HTML `fieldset` that pre-configures ARIA `role` for a radio group.

<Playground>
    {() => {
        const radio = useRadioState();

        return (
            <RadioGroup arial-label="fruits">
                <Radio {...radio} value="apple">Apple</Radio>
                <Radio {...radio} value="orange">Orange</Radio>
                <Radio {...radio} value="watermelon">Watermelon</Radio>
            </RadioGroup>
        );
    }}
</Playground>
