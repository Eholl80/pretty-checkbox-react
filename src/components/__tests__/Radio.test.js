import React from 'react';

import { render, cleanup, getByTestId, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import Radio from '../Radio';

afterEach(cleanup);

describe('Radio tests', function () {
    describe('Radio usage as a radio', function () {
        it('should behave as a radio', function () {
            const handleChange = jest.fn();
            const { container } = render(<Radio id="foo" onChange={handleChange} />);

            expect(getByTestId(container, 'pcr-input')).toHaveAttribute('type', 'radio');

            fireEvent.click(getByTestId(container, 'pcr-input'));

            expect(handleChange).toHaveBeenCalled();
            expect(handleChange).toHaveBeenCalledTimes(1);
            expect(getByTestId(container, 'pcr-input').checked).toBe(true);

            fireEvent.click(getByTestId(container, 'pcr-input'));

            expect(handleChange).toHaveBeenCalledTimes(1);
            expect(getByTestId(container, 'pcr-input').checked).toBe(true);
            expect(container).toMatchSnapshot();
        });

        it('should behave as a radio group', function () {
            const handleChange = jest.fn();

            const { container } = render(
                <>
                    <Radio id="demo" name="radio-1" value="r1" onChange={handleChange}>Hello there.</Radio>
                    <Radio id="demo" name="radio-1" value="r2" onChange={handleChange}>General Kenobi. You are a bold one.</Radio>
                    <Radio id="demo" name="radio-1" value="r3" onChange={handleChange}>Your move.</Radio>
                </>
            );

            const r1 = container.firstChild.firstChild;
            const r2 = container.firstChild.nextSibling.firstChild;
            const r3 = container.firstChild.nextSibling.nextSibling.firstChild;

            fireEvent.click(r1);

            expect(handleChange).toHaveBeenCalled();
            expect(handleChange).toHaveBeenCalledTimes(1);
            expect(r1.checked).toBe(true);

            fireEvent.click(r2);

            expect(handleChange).toHaveBeenCalledTimes(2);
            expect(r1.checked).toBe(false);
            expect(r2.checked).toBe(true);

            fireEvent.click(r3);

            expect(handleChange).toHaveBeenCalledTimes(3);
            expect(r1.checked).toBe(false);
            expect(r2.checked).toBe(false);
            expect(r3.checked).toBe(true);
            expect(container).toMatchSnapshot();
        });

        it('should render icons with the correct classNames', function () {
            let r1 = render(<Radio icon={<i className="mdi mdi-check" />}>Hello there.</Radio>).container;
            let r2 = render(<Radio svg={<i className="mdi mdi-check" />}>Hello there.</Radio>).container;
            let r3 = render(<Radio image={<i className="mdi mdi-check" />}>Hello there.</Radio>).container;

            expect(getByTestId(r1, 'pcr-wrapper')).toHaveClass('p-icon');
            expect(getByTestId(r2, 'pcr-wrapper')).toHaveClass('p-svg');
            expect(getByTestId(r3, 'pcr-wrapper')).toHaveClass('p-image');
        });
    });

    describe('Radio ref tests', function() {
        it('should accept custom refs', function () {
            const ref = React.createRef();

            render(<Radio ref={ref} />);
            expect(ref.current).toBeDefined();
            expect(ref.current).toBeInstanceOf(HTMLInputElement);
        });
    });
});
