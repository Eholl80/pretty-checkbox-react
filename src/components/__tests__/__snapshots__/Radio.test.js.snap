// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio tests Radio usage as a radio should behave as a radio 1`] = `
<div>
  <div
    class="pretty p-default p-round"
    data-testid="pcr-wrapper"
  >
    <input
      data-testid="pcr-input"
      id="foo"
      type="radio"
      value=""
    />
    <div
      class="state"
      data-testid="pcr-state"
    >
      <label
        for="foo"
      />
    </div>
  </div>
</div>
`;

exports[`Radio tests Radio usage as a radio should behave as a radio group 1`] = `
<div>
  <div
    class="pretty p-default p-round"
    data-testid="pcr-wrapper"
  >
    <input
      data-testid="pcr-input"
      id="demo"
      name="radio-1"
      type="radio"
      value="r1"
    />
    <div
      class="state"
      data-testid="pcr-state"
    >
      <label
        for="demo"
      >
        Hello there.
      </label>
    </div>
  </div>
  <div
    class="pretty p-default p-round"
    data-testid="pcr-wrapper"
  >
    <input
      data-testid="pcr-input"
      id="demo"
      name="radio-1"
      type="radio"
      value="r2"
    />
    <div
      class="state"
      data-testid="pcr-state"
    >
      <label
        for="demo"
      >
        General Kenobi. You are a bold one.
      </label>
    </div>
  </div>
  <div
    class="pretty p-default p-round"
    data-testid="pcr-wrapper"
  >
    <input
      data-testid="pcr-input"
      id="demo"
      name="radio-1"
      type="radio"
      value="r3"
    />
    <div
      class="state"
      data-testid="pcr-state"
    >
      <label
        for="demo"
      >
        Your move.
      </label>
    </div>
  </div>
</div>
`;
